File: nodejs/MinioNodejs.Services/src/services/MinioService.ts
import * as Minio from "minio";
import { Readable } from "stream";
import { IMinioService } from "../interfaces/IMinioService";

export class MinioService implements IMinioService {
  constructor(private readonly client: Minio.Client) {}

  async createBucket(bucketName: string): Promise<void> {
    const exists = await this.client.bucketExists(bucketName);
    if (!exists) {
      await this.client.makeBucket(bucketName);
    }
  }

  async listBuckets(): Promise<string[]> {
    const buckets = await this.client.listBuckets();
    return buckets.map((b) => b.name);
  }

  async bucketExists(bucketName: string): Promise<boolean> {
    return this.client.bucketExists(bucketName);
  }

  async deleteBucket(bucketName: string): Promise<void> {
    await this.client.removeBucket(bucketName);
  }

  async uploadFile(
    bucketName: string,
    objectName: string,
    stream: Readable,
    size: number,
    contentType: string
  ): Promise<void> {
    await this.client.putObject(bucketName, objectName, stream, size, {
      "Content-Type": contentType,
    });
  }

  async downloadFile(bucketName: string, objectName: string): Promise<Readable> {
    return this.client.getObject(bucketName, objectName);
  }
}

File: java/MinioJava.Services/src/main/java/com/example/minio/services/MinioService.java
package com.example.minio.services;

import io.minio.*;
import io.minio.messages.Bucket;

import java.io.InputStream;
import java.util.List;

public class MinioService implements IMinioService {

    private final MinioClient client;

    public MinioService(MinioClient client) {
        this.client = client;
    }

    @Override
    public void createBucket(String bucketName) throws Exception {
        boolean exists = client.bucketExists(BucketExistsArgs.builder().bucket(bucketName).build());
        if (!exists) {
            client.makeBucket(MakeBucketArgs.builder().bucket(bucketName).build());
        }
    }

    @Override
    public List<String> listBuckets() throws Exception {
        return client.listBuckets().stream().map(Bucket::name).toList();
    }

    @Override
    public boolean bucketExists(String bucketName) throws Exception {
        return client.bucketExists(BucketExistsArgs.builder().bucket(bucketName).build());
    }

    @Override
    public void deleteBucket(String bucketName) throws Exception {
        client.removeBucket(RemoveBucketArgs.builder().bucket(bucketName).build());
    }

    @Override
    public void uploadFile(String bucketName, String objectName, InputStream stream, long size, String contentType) throws Exception {
        client.putObject(PutObjectArgs.builder()
                .bucket(bucketName)
                .object(objectName)
                .stream(stream, size, -1)
                .contentType(contentType)
                .build());
    }

    @Override
    public InputStream downloadFile(String bucketName, String objectName) throws Exception {
        return client.getObject(GetObjectArgs.builder()
                .bucket(bucketName)
                .object(objectName)
                .build());
    }
}
